cmd/zdb/zdb.c

```diff
@@ -1557,6 +1562,64 @@ blkid2offset(const dnode_phys_t *dnp, const blkptr_t *bp,
 	    dnp->dn_datablkszsec << SPA_MINBLOCKSHIFT);
 }
 
+static void
+snprintf_zstd_header(spa_t *spa, char *blkbuf, size_t buflen,
+    const blkptr_t *bp)
+{
+	abd_t *pabd;
+	void *buf;
+	zio_t *zio;
+	zfs_zstdhdr_t *zstd_hdr;
+	int error;
+
+	if (BP_GET_COMPRESS(bp) != ZIO_COMPRESS_ZSTD)
+		return;
+
+	if (BP_IS_HOLE(bp))
+		return;
+
+	if (BP_IS_EMBEDDED(bp)) {
+		buf = malloc(SPA_MAXBLOCKSIZE);
+		if (buf == NULL) {
+			(void) fprintf(stderr, "out of memory\n");
+			exit(1);
+		}
+		decode_embedded_bp_compressed(bp, buf);
+		zstd_hdr = (zfs_zstdhdr_t *)buf;
+		zstd_hdr->c_len = BE_32(zstd_hdr->c_len);
+		zstd_hdr->raw_version_level = BE_32(zstd_hdr->raw_version_level);
```

Comment:
i hope this buffer is not reused a second time. you are changing the content here while doing endian conversation

Reply:
what's the problem? buf is a local buffer and is a copy

Reply:
I might be worth adding a comment here, as having done the byte swap, if you actually feed this to zio_decompress it will byte swap it again, and bad things will happen.

Reply:
@c0d3z3r0 my fault

Reply:
I have a fix pushed already, it was a good catch

Reply:
oh okay. then it not my fault :-)